tests/integration/telemetry/stats/prometheus/wasm/imagepullpolicy_test.go

```diff
@@ -0,0 +1,251 @@
+//go:build integ
+// +build integ
+
+// Copyright Istio Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wasm
+
+import (
+	"errors"
+	"fmt"
+	"testing"
+	"time"
+
+	"istio.io/istio/pkg/test/framework"
+	"istio.io/istio/pkg/test/framework/components/echo"
+	"istio.io/istio/pkg/test/scopes"
+	"istio.io/istio/pkg/test/util/retry"
+	common "istio.io/istio/tests/integration/telemetry/stats/prometheus"
+)
+
+const (
+	imageName      = "istio-testing/wasm/header-injector"
+	injectedHeader = "x-resp-injection"
+	wasmConfigFile = "testdata/wasm-filter.yaml"
+)
+
+type wasmTestConfigs struct {
+	desc            string
+	name            string
+	policy          string
+	tag             string
+	upstreamVersion string
+	expectedVersion string
+}
+
+var generation = 0
+
+func mapTagToVersionOrFail(t framework.TestContext, tag, version string) {
+	t.Helper()
+	if err := registry.SetupTagMap(map[string]string{
+		imageName + ":" + tag: version,
+	}); err != nil {
+		t.Fatalf("failed to setup the tag map: %v", err)
+	}
+}
+
+func applyAndTestWasm(t framework.TestContext, c wasmTestConfigs) {
+	t.Helper()
+	defer func() {
+		generation++
+	}()
+	scopes.Framework.Infof(c.desc)
+	mapTagToVersionOrFail(t, c.tag, c.upstreamVersion)
+	if err := installWasmExtension(t, c.name, c.tag, c.policy, fmt.Sprintf("g-%q", generation)); err != nil {
+		t.Fatalf("failed to install WasmPlugin: %v", err)
+	}
+	retry.UntilSuccessOrFail(t, func() error {
```

Comment:
Would prefer to use `Check: check.ResponseHeader(injectedHeader, c.expectedVersion),` in the CallOptions

Something like

```go

	cltInstance := common.GetClientInstances()[0]

	httpOpts := echo.CallOptions{
		To: common.GetTarget(),
		Port: echo.Port{
			Name: "http",
		},
		HTTP: echo.HTTP{
			Path:   "/path",
			Method: "GET",
		},
		Count: 1,
		Check: check.ResponseHeader(injectedHeader, c.expectedVersion),
	}
  cltInstance.CallOrFail(t, httpOpts)
```

Then there is far less error handling done here and we get the more robust common code

Reply:
Thank you for the detailed comment. Now the code is much clearer.